README.rst
pointhq
=======

pointhq.com API client.

Installation
------------

Install ``pointhq`` with pip::

    $ pip install pointhq

It will also install ``httplib2`` library.

If you use Python 2.5, ``simplejson`` is required::

    $ pip install simplejson

Usage example
-------------

1. Create new ``pointhq.Point`` object::

    import pointhq
    point = Point(username='john@example.com', apitoken='secret-key')

2. Play with zones::

    zones = point.zones.retrieve()
    new_zone = point.zones.create(name='example.com')

    zone = point.zones(1).retrieve()
    point.zones(1).update(group='Clients')
    point.zones(1).delete()

3. Play with zone records::

    zone_records = point.zones(1).records.retrieve()
    new_record = point.zones(1).records.create(name='example.com.', data='123.45.67.89', record_type='A')

    zone_record = point.zones(1).records(1).retrieve
    point.zones(1).records(1).update(data='234.56.78.90')
    point.zones(1).records(1).delete()

Contributing
------------

Feel free to fork, send pull requests or report bugs and issues on github.


HISTORY.rst
History
=======

0.1.1 (2011-10-11)
------------------

- Fix ``request`` function

0.1 (2011-10-10)
----------------

- Initial release


tests/test_zones.py
from __future__ import with_statement

import unittest2
from mock import Mock

from pointhq import Point
from pointhq import exceptions
from pointhq.helpers import Response


class ZoneTests(unittest2.TestCase):

    def setUp(self):
        self.auth = ('john@example.com', 'secret-key')
        self.request = Mock()
        self.point = Point(self.auth[0], self.auth[1], self.request)

    def test_collection_retrieve(self):
        self.request.return_value = Response(200, '[{"zone": {"id": 1}}]')
        zone = self.point.zones.retrieve()
        self.request.assert_called_once_with('get', '/zones', self.auth, None)
        self.assertEqual(zone, [{'zone': {'id': 1}}])

    def test_collection_retrieve_failure(self):
        self.request.return_value = Response(201)
        with self.assertRaises(exceptions.UnknownError):
            self.point.zones.retrieve()

    def test_collection_create(self):
        self.request.return_value = Response(201, '{"zone": {"id": 1}}')
        zone = self.point.zones.create(name='example.com')
        self.request.assert_called_once_with('post', '/zones', self.auth, {
            'zone': {
                'name': 'example.com',
            }
        })
        self.assertEqual(zone, {'zone': {'id': 1}})

    def test_collection_create_failure(self):
        self.request.return_value = Response(422, '["error message"]')
        with self.assertRaises(exceptions.UnprocessableEntityError):
            self.point.zones.create(name='example.com')

    def test_collection_create_unknown_failure(self):
        self.request.return_value = Response(200)
        with self.assertRaises(exceptions.UnknownError):
            self.point.zones.create(name='example.com')

    def test_member_retrieve(self):
        self.request.return_value = Response(200, '{"zone": {"id": 1}}')
        zone = self.point.zones(1).retrieve()
        self.request.assert_called_once_with('get', '/zones/1', self.auth, None)
        self.assertEqual(zone, {'zone': {'id': 1}})

    def test_member_retrieve_failure(self):
        self.request.return_value = Response(404)
        with self.assertRaises(exceptions.NotFoundError):
            self.point.zones(1).retrieve()

    def test_member_retrieve_unknown_failure(self):
        self.request.return_value = Response(201)
        with self.assertRaises(exceptions.UnknownError):
            self.point.zones(1).retrieve()

    def test_member_update(self):
        self.request.return_value = Response(200)
        self.assertIsNone(self.point.zones(1).update(group='clients'))
        self.request.assert_called_once_with('put', '/zones/1', self.auth, {
            'zone': {
                'group': 'clients',
            }
        })

    def test_member_update_failure(self):
        self.request.return_value = Response(422, '["error message"]')
        with self.assertRaises(exceptions.UnprocessableEntityError):
            self.point.zones(1).update(group='clients')

    def test_member_update_unknown_failure(self):
        self.request.return_value = Response(404)
        with self.assertRaises(exceptions.UnknownError):
            self.point.zones(1).update(group='clients')

    def test_member_delete(self):
        self.request.return_value = Response(200)
        self.assertIsNone(self.point.zones(1).delete())
        self.request.assert_called_once_with('delete', '/zones/1', self.auth, None)

    def test_member_delete_failure(self):
        self.request.return_value = Response(409, '["error message"]')
        with self.assertRaises(exceptions.ConflictError):
            self.point.zones(1).delete()

    def test_member_delete_unknown_failure(self):
        self.request.return_value = Response(404)
        with self.assertRaises(exceptions.UnknownError):
            self.point.zones(1).delete()


tests/test_request.py
import unittest2
import mock
from pointhq.helpers import request


class RequestTests(unittest2.TestCase):

    @mock.patch('httplib2.Http')
    def test_request(self, Http):
        http_response = mock.Mock()
        instance = Http.return_value
        instance.request.return_value = http_response, 'content'

        request('get', '/', ('john', 'secret-key'))
        Http.assert_called_once_with(timeout=10)
        self.assertTrue(instance.request.called)


tests/test_zone_records.py
from __future__ import with_statement

import unittest2
from mock import Mock

from pointhq import Point
from pointhq import exceptions
from pointhq.helpers import Response


class ZoneRecordTests(unittest2.TestCase):

    def setUp(self):
        self.auth = ('john@example.com', 'secret-key')
        self.request = Mock()
        self.point = Point(self.auth[0], self.auth[1], self.request)

    def test_collection_retrieve(self):
        self.request.return_value = Response(200, '[{"zone_record": {"id": 1}}]')
        zone_record = self.point.zones(1).records.retrieve()
        self.request.assert_called_once_with('get', '/zones/1/records', self.auth, None)
        self.assertEqual(zone_record, [{'zone_record': {'id': 1}}])

    def test_collection_retrieve_unknown_failure(self):
        self.request.return_value = Response(404)
        with self.assertRaises(exceptions.UnknownError):
            self.point.zones(1).records.retrieve()

    def test_collection_create(self):
        self.request.return_value = Response(201, '{"zone_record": {"id": 1}}')
        zone_record = self.point.zones(1).records.create(name='example.com.', data='123.45.67.89')
        self.request.assert_called_once_with('post', '/zones/1/records', self.auth, {
            'zone_record': {
                'name': 'example.com.',
                'data': '123.45.67.89',
            }
        })
        self.assertEqual(zone_record, {'zone_record': {'id': 1}})

    def test_collection_create_failure(self):
        self.request.return_value = Response(422, '["error message"]')
        with self.assertRaises(exceptions.UnprocessableEntityError):
            self.point.zones(1).records.create(name='example.com.')

    def test_collection_create_unknown_failure(self):
        self.request.return_value = Response(404)
        with self.assertRaises(exceptions.UnknownError):
            self.point.zones(1).records.create(name='example.com.')

    def test_member_retrieve(self):
        self.request.return_value = Response(200, '{"zone_record": {"id": 1}}')
        zone_record = self.point.zones(1).records(1).retrieve()
        self.request.assert_called_once_with('get', '/zones/1/records/1', self.auth, None)
        self.assertEqual(zone_record, {'zone_record': {'id': 1}})

    def test_member_retrieve_failure(self):
        self.request.return_value = Response(404)
        with self.assertRaises(exceptions.NotFoundError):
            self.point.zones(1).records(1).retrieve()

    def test_member_retrieve_unknown_failure(self):
        self.request.return_value = Response(500)
        with self.assertRaises(exceptions.UnknownError):
            self.point.zones(1).records(1).retrieve()

    def test_member_update(self):
        self.request.return_value = Response(200)
        self.assertIsNone(self.point.zones(1).records(1).update(data='123.45.67.89'))
        self.request.assert_called_once_with('put', '/zones/1/records/1', self.auth, {
            'zone_record': {
                'data': '123.45.67.89',
            }
        })

    def test_member_update_failure(self):
        self.request.return_value = Response(422, '["error message"]')
        with self.assertRaises(exceptions.UnprocessableEntityError):
            self.point.zones(1).records(1).update(data='123.45.67.89')

    def test_member_update_unknown_failure(self):
        self.request.return_value = Response(500)
        with self.assertRaises(exceptions.UnknownError):
            self.point.zones(1).records(1).update(data='123.45.67.89')

    def test_member_delete(self):
        self.request.return_value = Response(200)
        self.assertIsNone(self.point.zones(1).records(1).delete())
        self.request.assert_called_once_with('delete', '/zones/1/records/1', self.auth, None)

    def test_member_delete_failure(self):
        self.request.return_value = Response(409)
        with self.assertRaises(exceptions.ConflictError):
            self.point.zones(1).records(1).delete()

    def test_member_delete_unknown_failure(self):
        self.request.return_value = Response(500)
        with self.assertRaises(exceptions.UnknownError):
            self.point.zones(1).records(1).delete()


pointhq/api.py
from . import base
from . import exceptions
from . import helpers


class Point(object):

    def __init__(self, username, apitoken, request_func=helpers.request):
        self.auth = (username, apitoken)
        self.request_func = request_func

    @property
    def zones(self):
        return ZoneCollection(request_func=self.request)

    def request(self, method, url, data=None):
        return self.request_func(method, url, self.auth, data)


class Zone(base.BaseMember):

    name = 'zone'
    path_pattern = '/zones/%(id)s'

    @property
    def records(self):
        return ZoneRecordCollection(
            zone_id = self.kwargs['id'],
            request_func = self.kwargs['request_func'],
        )


class ZoneCollection(base.BaseCollection):

    name = 'zone'
    path_pattern = '/zones'
    member_class = Zone


class ZoneRecord(base.BaseMember):

    name = 'zone_record'
    path_pattern = '/zones/%(zone_id)s/records/%(id)s'


class ZoneRecordCollection(base.BaseCollection):

    name = 'zone_record'
    path_pattern = '/zones/%(zone_id)s/records'
    member_class = ZoneRecord


pointhq/base.py
try:
    import simplejson as json
except ImportError:
    import json

from . import exceptions


class BaseResourse(object):

    name = None
    path_pattern = None

    def __init__(self, **kwargs):
        self.kwargs = kwargs

    @property
    def path(self):
        return self.path_pattern % self.kwargs

    def request(self, method, data=None):
        if data is not None:
            data = {self.name: data}
        return self.kwargs['request_func'](method, self.path, data)


class BaseMember(BaseResourse):

    def retrieve(self):
        response = self.request('get')
        if response.status == 200:
            return json.loads(response.content)
        if response.status == 404:
            raise exceptions.NotFoundError
        raise exceptions.UnknownError(response)

    def update(self, **kwargs):
        response = self.request('put', kwargs)
        if response.status == 200:
            return
        if response.status == 422:
            raise exceptions.UnprocessableEntityError(json.loads(response.content))
        raise exceptions.UnknownError(response)

    def delete(self):
        response = self.request('delete')
        if response.status == 200:
            return
        if response.status == 409:
            raise exceptions.ConflictError
        raise exceptions.UnknownError(response)


class BaseCollection(BaseResourse):

    member_class = None

    def __call__(self, id):
        kwargs = self.kwargs.copy()
        kwargs['id'] = id
        return self.member_class(**kwargs)

    def retrieve(self):
        response = self.request('get')
        if response.status == 200:
            return json.loads(response.content)
        raise exceptions.UnknownError(response)

    def create(self, **kwargs):
        response = self.request('post', kwargs)
        if response.status == 201:
            return json.loads(response.content)
        if response.status == 422:
            raise exceptions.UnprocessableEntityError(json.loads(response.content))
        raise exceptions.UnknownError(response)
